### PR TITLE
Ft verification email validation

### PR DESCRIPTION
**Pull request by kennyg37**
#### What does this PR do?
configures an automatic email as soon as a user signs up and validates the user using a jwt token connected to 
a link in the email
#### Describe your changes
added a controller file to handle the token verification after the link is clicked
#### PR Type
What kind of change does this PR introduce?
```
[ ] Bugfix
[x] Feature
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Tests
[ ] Other
```

#### What are the relevant pivotal tracker stories?
[Pivotal Tracker Link](https://xsolve.atlassian.net/browse/PROJECT_KEY)

#### Screenshots (if is appropriate)

### CODE REVIEW

**Review by mugishaj092 — CHANGES_REQUESTED** [review]
generateVerificationToken function it's not necessary you can reuse generateToken and add tests

**Review by GabrielDushime — APPROVED** [review]

**Review by mutsinziisaac — DISMISSED** [review]

**Review by GabrielDushime — DISMISSED** [review]
Good

**Review by mugishaj092 — DISMISSED** [review]

**Review by SHYAKA-Aime — APPROVED** [review]

File: .env.example [hunk 1]
```diff
@@ -18,4 +18,6 @@ PRO_DB_PASS =
 PRO_DB_HOST = 
 
 EMAIL_USER =
-EMAIL_PASS = 
\ No newline at end of file
+EMAIL_PASS = 
+
+FRONTEND_URL =
```
Comment (didasmbalanya) [inline .env.example:23]: Add empty line at the end of each file

File: src/controllers/user.controller.ts [hunk 2]
```diff
@@ -29,8 +26,12 @@ export const userSignup = async (req: Request, res: Response) => {
     }
 
     const createdUser = await UserService.register(user);
-    const token = await generateToken(createdUser);
-    sendVerificationEmail(user.email, subject, text, html);
+    const token = await generateToken(createdUser, "1h");
+
+    const verificationLink = `${process.env.FRONTEND_URL}/api/users/verify-email?token=${token}`;
+    const text = `Please verify your email by clicking on the following link:${verificationLink}`;
+    const html = `<p>Please verify your email by clicking on the following link:</p><a href="${verificationLink}">Verify Email</a>`;
```
Comment (didasmbalanya) [inline src/controllers/user.controller.ts:33]: Move this to your email sending logic

File: src/controllers/verifyUser.controller.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  const { token } = req.query;
+  if (!token) {
+    return res.status(400).json({ message: "Invalid token" });
+  }
+  try {
+    const decoded: any = jwt.verify(token as string, process.env.JWT_SECRET || "");
+    const user = await UserService.getUserByEmail(decoded.email);
+    if (!user) {
+      return res.status(404).json({ message: "User not found" });
+    }
```
Comment (didasmbalanya) [inline src/controllers/verifyUser.controller.ts:19]: Move this to a middleware and make it reuseable

File: src/controllers/verifyUser.controller.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    user.verified = true;
+    await UserService.updateUser(user);
+    const authToken = await generateToken(user, "24h");
```
Comment (didasmbalanya) [inline src/controllers/verifyUser.controller.ts:22]: Make the token validity period configurable. i.e an env

File: src/controllers/verifyUser.controller.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+      status: "success",
+      message: "Email verified successfully",
+      token: authToken,
+    }
+  );
+  } catch (error) {
+    if (error instanceof jwt.JsonWebTokenError) {
+      return res.status(400).json({ message: "Invalid token" });
+    }
+    return res.status(500).json({ message: "Invalid token" });
```
Comment (didasmbalanya) [inline src/controllers/verifyUser.controller.ts:33]: Handle better

File: src/utils/tokenGenerator.utils.ts [hunk 6]
```diff
@@ -4,7 +4,7 @@ import dotenv from "dotenv";
 
 dotenv.config();
 
-const generateToken = async (user: UserAttributes) => {
+const generateToken = async (user: UserAttributes, expiresIn: string = "30d") => {
```
Comment (didasmbalanya) [inline src/utils/tokenGenerator.utils.ts:7]: Make the expiry cofigurable

**Review by didasmbalanya — CHANGES_REQUESTED** [review]
Good job so far but there are a few changes I have pointed out. please look into them

**Pull request closed without merging**